core/schemas/responses.go

```
@@ -1823,16 +1862,58 @@ func (action ResponsesToolMessageActionStruct) MarshalJSON() ([]byte, error) {
+
+	// First, peek at the type field to determine which object variant to unmarshal.
 	var typeStruct struct {
 		Type string `json:"type"`
 	}
 	if err := Unmarshal(data, &typeStruct); err != nil {
-		return fmt.Errorf("failed to peek at type field: %w", err)
+		// Preserve a provider-native action shape that is not an object with a
+		// string discriminator. This keeps the surrounding Responses item
+		// decodable while retaining bytes for a later round trip.
+		action.Raw = append([]byte(nil), trimmed...)
+		return nil
```

golangci-lint (2.12.2)

[error] error is not nil (line 1911) but it returns nil

(nilerr)